components/transaction.tsx

```diff
@@ -176,7 +173,70 @@ const NewTransaction = ({ balance }: NewTransactionProps) => {
         _quantity && assets.set(id, _quantity - quantity)
       })
       return { lovelace, assets }
-    }, balance.value)
+    }, getBalance(utxos))
+
+
+  const buildTransaction = (): { type: 'ok', transaction: TransactionBody } | { type: 'error', message: string } => {
+    try {
+      const txBuilder = cardano.createTxBuilder(protocolParameters)
+      const { Address, AssetName, BigNum, TransactionOutputBuilder, MultiAsset, ScriptHash } = cardano.lib
+      recipients.forEach((recipient) => {
+        const address = Address.from_bech32(recipient.address)
+        const txOutputBuilder = TransactionOutputBuilder
+          .new()
+          .with_address(address)
+          .next()
+        const { lovelace, assets } = recipient.value
+        const value = cardano.lib.Value.new(BigNum.from_str(lovelace.toString()))
+        if (assets.size > 0) {
+          const multiAsset = MultiAsset.new()
+          assets.forEach((quantity, id, _) => {
+            const policyId = ScriptHash.from_bytes(Buffer.from(getPolicyId(id), 'hex'))
+            const assetName = AssetName.new(Buffer.from(getAssetName(id), 'hex'))
+            const value = BigNum.from_str(quantity.toString())
+            multiAsset.set_asset(policyId, assetName, value)
+          })
+          value.set_multiasset(multiAsset)
+        }
+        const txOutput = txOutputBuilder.with_value(value).build()
+        txBuilder.add_output(txOutput)
+      })
+
+      const { TransactionInput, TransactionHash, TransactionOutput, TransactionUnspentOutput, TransactionUnspentOutputs } = cardano.lib
+      const utxosSet = TransactionUnspentOutputs.new()
+      const utxosAddress = Address.from_bech32(senderAddress)
+      utxos.forEach((utxo) => {
+        const { txHash, index, lovelace, assets } = utxo
+        const value = cardano.lib.Value.new(BigNum.from_str(lovelace.toString()))
+        if (assets.length > 0) {
+          const multiAsset = MultiAsset.new()
+          assets.forEach((asset) => {
+            const policyId = ScriptHash.from_bytes(Buffer.from(asset.policyId, 'hex'))
+            const assetName = AssetName.new(Buffer.from(asset.assetName, 'hex'))
+            const quantity = BigNum.from_str(asset.quantity.toString())
+            multiAsset.set_asset(policyId, assetName, quantity)
+          })
+          value.set_multiasset(multiAsset)
+        }
+        const txUnspentOutput = TransactionUnspentOutput.new(
+          TransactionInput.new(TransactionHash.from_bytes(Buffer.from(txHash, 'hex')), index),
+          TransactionOutput.new(utxosAddress, value)
+        )
+        utxosSet.add(txUnspentOutput)
+      })
+      txBuilder.add_inputs_from(utxosSet, cardano.lib.CoinSelectionStrategyCIP2.LargestFirstMultiAsset)
```

Comment:
this didn't work for me , I used `CoinSelectionStrategyCIP2.RandomImproveMultiAsset`